src/sql/base/browser/ui/table/treeGrid.ts

```diff
@@ -0,0 +1,176 @@
+
+import 'vs/css!./media/slick.grid';
+
+import { FilterableColumn, ITableConfiguration } from 'sql/base/browser/ui/table/interfaces';
+import { Table } from 'sql/base/browser/ui/table/table';
+import { treeGridExpandableColumnFormatter } from 'sql/base/browser/ui/table/formatters';
+import { IDisposableDataProvider } from 'sql/base/common/dataProvider';
+import { generateUuid } from 'vs/base/common/uuid';
+import { CellValueGetter, defaultCellValueGetter, defaultFilter, TableDataView } from 'sql/base/browser/ui/table/tableDataView';
+import { AsyncDataProvider } from 'sql/base/browser/ui/table/asyncDataView';
+import { isArray } from 'vs/base/common/types';
+import { StandardKeyboardEvent } from 'vs/base/browser/keyboardEvent';
+import { KeyCode } from 'vs/base/common/keyCodes';
+
+
+function defaultTreeGridFilter<T extends Slick.SlickData>(data: T[], columns: FilterableColumn<T>[], cellValueGetter: CellValueGetter = defaultCellValueGetter): T[] {
+	let filteredData = defaultFilter(data, columns, cellValueGetter);
+
+	// filtering out rows which have parent/grandparents collapsed.
+	filteredData = filteredData.filter((item) => {
+		let parent = data[item.parent];
+		while (parent) {
+			if (!parent.expanded) {
+				return false;
+			}
+			parent = data[parent.parent];
+		}
+		return true;
+	});
+
+	return filteredData;
+}
+
+
+export class TreeGrid<T extends Slick.SlickData> extends Table<T> {
+	constructor(parent: HTMLElement, configuration?: ITableConfiguration<T>, options?: Slick.GridOptions<T>) {
+		super(parent, configuration, options);
+
+		if (!configuration || !configuration.dataProvider || isArray(configuration.dataProvider)) {
+			this._data = new TableDataView<T>(configuration && configuration.dataProvider as Array<T>,
+				undefined,
+				undefined,
+				defaultTreeGridFilter,
+				undefined);
+		} else {
+			this._data = configuration.dataProvider;
+		}
+
+		this._grid.onClick.subscribe((e, data) => {
+			this.setCellExpandedState(data.row, data.cell);
+			return false;
+		});
+
+		this._grid.onKeyDown.subscribe((e, data) => {
+			const keyboardEvent = (<any>e).originalEvent;
+			if (keyboardEvent instanceof KeyboardEvent) {
+				let event = new StandardKeyboardEvent(keyboardEvent);
+				if (event.keyCode === KeyCode.Enter) {
+					// toggle the collapsed state of the row
+					this.setCellExpandedState(data.row, data.cell);
+				} else if (event.keyCode === KeyCode.LeftArrow) {
+					// Left arrow on first cell of the expanded row collapses it
+					if (data.cell === 0) {
+						this.setCellExpandedState(data.row, this.expandableColumnIndex(), false);
+					}
+				} else if (event.keyCode === KeyCode.RightArrow) {
+					// Right arrow on last cell of the collapsed row expands it.
+					if (data.cell === (this._grid.getColumns().length - 1)) {
+						this.setCellExpandedState(data.row, this.expandableColumnIndex(), true);
+					}
+				}
+			}
+			return false;
```

Comment:
what is this for?

Reply:
Since these are Jquery events, returning false does preventDefault and stopPropogation.